### PR TITLE
B parking Skimming code update - backport 102X

### PR DESCRIPTION
**Pull request by gkaratha**
Dear all,

Please review this PR. Is a backport of  #25984 as was requested. Nothing is different in terms of code ( with the small exception that ConversionTools interface is different in 10_2X than 10_5X, so i adjusted it). The rate has been measured and is 5.35% as in the  #25984. Thanks

Best regards,
George

### CODE REVIEW

**Comment by gkaratha** [comment]
backport of https://github.com/cms-sw/cmssw/pull/25984 . @kfjack @bainbrid @mverzett

**Comment by boudoul** [comment]
Please @gkaratha , provide an explicit title. For backport, in general providing the same title as the corresponding one of master plus indicating that it is a backport is fine - 
Now I see that your master PR is named 'Solving the conflicts' which is definitely not appropriate, so please provide an explicit title of your master PR in addition to this one- Thank you.

**Comment by gkaratha** [comment]
Dear @boudoul ,

Thanks for the comment, I did not know. I changed it. 

Best regards,
George

**Comment by prebello** [comment]
please test

**Comment by boudoul** [comment]
Hello @gkaratha , now that you gave a more appropriate title on the master PR  #25984  , you could edit this tiltle and make it also more appropriate [this is essentially what I was asking earlier, maybe not clear enough]  
for example it could be  'B parking Skimming code update - backport 102X'

**Comment by gkaratha** [comment]
Dear @boudoul ,

Yes you are right! I changed the title. Thanks for the suggestion, is better from what i had.

Best regards,
George

**Comment by gkaratha** [comment]
Dear all,

I have updated the WP. The new one has 6.5% rate while keeping a bit more than 29% in signal efficiency. Thanks.

Best regards,
George

**Comment by prebello** [comment]
please test

**Comment by prebello** [comment]
Hi @kfjack please take a look too.

File: Configuration/Skimming/src/LeptonSkimming.cc [hunk 1]
```diff
@@ -0,0 +1,487 @@
...
+   }
+  
+
+  //Save electrons we want to skim
+  ElTracks.clear();
+  if (!SkimOnlyMuons){
+    unsigned int count_el=-1;
+   for(const reco::GsfElectron &el : *electrons){  
+      count_el++;
+      bool passConvVeto = !ConversionTools::hasMatchedConversion(*(&el), conversions, theBeamSpot->position());
```
Comment (fabiocos) [inline Configuration/Skimming/src/LeptonSkimming.cc]: @gkaratha this line differs from the master version, where you use "*conversions", why? Here you are passing directly the Handle, I would say that the master version is the correct one

**Comment by bainbrid** [comment]
Hi Fabio,
The interface of the static ConversionTools::hasMatchedConversion method has changed b/w 10_2_X and master, hence the difference you refer to. 
Rob

**Comment by fabiocos** [comment]
@bainbrid ok, the change is in https://github.com/cms-sw/cmssw/commit/c5bc1c243e6b362e748a87fc6692d6d6956ee810#diff-bdfe7dd18f66eedb798c424905d1f5b7 where the Handle has been replaced by the collection itself in 10_4_X.

**Comment by kfjack** [comment]
Hi all, I compared the main code and implementation with the master, and I think this back port is fine in terms of skim part. Probably we should also backport the RelVal.
@gkaratha do you want to add them into this PR? e.g. add these modifications I introduced to PR #25892. If not, I can do that for you by restarting a new PR probably.

**Comment by gkaratha** [comment]
Dear @kfjack,

I think I can do it with proper instructions. Could tell me how exactly?Thanks.

Best regards,
George

**Comment by kfjack** [comment]
Hi @gkaratha
Sure, I think Configuration/PyReleaseValidation package is needed, and you can just insert the lines I put into 
Configuration/PyReleaseValidation/python/relval_standard.py
and 
Configuration/PyReleaseValidation/python/relval_steps.py
in PR #25892. 

You can test the workflow offline by
runTheMatrix.py --what standard -l 136.898 --command=' -n 10'

and see if it works. If yes, you can commit the updates here. If you are running into any trouble, please let me know.

**Comment by gkaratha** [comment]
Dear @kfjack, 
I'm happy to do it if you can give me insturctions. Otherwise, I'd be very grateful if you are able to do it. I think @bainbrid is hoping to get this PR merged quickly... Thanks.

Best, george

**Comment by gkaratha** [comment]
Hi all,

Ok I did not receive @kfjack 's response before i send mine... Ok i will try today and let you know. I think what matters most is not to waste time on that. Thanks

Best regards,
George

**Comment by fabiocos** [comment]
@kfjack @gkaratha please check #26035, I think that this should implement the checks discussed. If this is ok I would move forward with this PR, and backport that as soon as we converge in master

**Comment by gkaratha** [comment]
Dear @kfjack ,

I followed your instructions. Unfortunately we have a problem at step 3[*]. From the error i see that the code cannot find lowPtElectrons, which is the input. Are you sure that are contained in the relval file? Thanks.

Best,
George


[*]----- Begin Fatal Exception 01-Mar-2019 01:00:58 CET-----------------------
An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 317661 lumi: 393 event: 560561156 stream: 0
   [1] Running path 'SkimBParkPath'
   [2] Calling method for module LeptonSkimming/'SkimBPark'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: std::vector<reco::GsfElectron>
Looking for module label: lowPtGsfElectrons
Looking for productInstanceName:

**Comment by kfjack** [comment]
Hi @gkaratha please also add the updates in PR #26035 which modify the corresponding workflow.
If it's still not working I can dig into it later today.

Jack

**Comment by bainbrid** [comment]
Hi @kfjack if you have time, please could you could take a look? We can certainly try, but neither of us are knowledgeable about the details concerning RelVal workflows. Your help would be freestyle appreciated!

**Comment by kfjack** [comment]
Hi @bainbrid, @gkaratha, I have created a new PR #26038 which includes all the updates from this PR and #25892 and #26035.

**Comment by bainbrid** [comment]
@kfjack many thanks!! Shall we close this PR?

**Comment by gkaratha** [comment]
Dear @kfjack , @bainbrid ,

Thanks for the help.


Best,
George

**Comment by kfjack** [comment]
Hi @bainbrid, yes, once the new PR is tested, I think we can close this one.

**Comment by fabiocos** [comment]
@gkaratha #26038 has been merged, so this PR can be closed

**Comment by gkaratha** [comment]
Dear @fabiocos ,

Ok sure! I am doing it now. Thank.

Best regards,
George

**Pull request closed without merging**